src/ristretto/pedersen/extended_commitment_factory.rs

```diff
@@ -0,0 +1,451 @@
+// INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+// DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+// USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+use bulletproofs_plus::{generators::pedersen_gens::ExtensionDegree, PedersenGens};
+use curve25519_dalek::{ristretto::RistrettoPoint, scalar::Scalar, traits::Identity};
+
+use crate::{
+    commitment::{ExtendedHomomorphicCommitmentFactory, HomomorphicCommitment},
+    errors::RangeProofError,
+    ristretto::{
+        constants::{RISTRETTO_NUMS_POINTS, RISTRETTO_NUMS_POINTS_COMPRESSED},
+        pedersen::{
+            PedersenCommitment,
+            RISTRETTO_PEDERSEN_G,
+            RISTRETTO_PEDERSEN_G_COMPRESSED,
+            RISTRETTO_PEDERSEN_H,
+            RISTRETTO_PEDERSEN_H_COMPRESSED,
+        },
+        RistrettoPublicKey,
+        RistrettoSecretKey,
+    },
+};
+
+/// Generates extended Pederson commitments `sum(k_i.G_i) + v.H` using the provided base
+/// [RistrettoPoints](curve25519_dalek::ristretto::RistrettoPoints).
+/// Notes:
+///  - Homomorphism with public key only holds for extended commitments with `ExtensionDegree::Zero`
+#[derive(Debug, PartialEq, Clone)]
+pub struct ExtendedPedersenCommitmentFactory(pub(crate) PedersenGens<RistrettoPoint>);
+
+impl ExtendedPedersenCommitmentFactory {
+    /// Create a new Extended Pedersen Ristretto Commitment factory for the required extension degree using
+    /// pre-calculated compressed constants - we only hold references to the static generator points.
+    pub fn new_with_extension_degree(extension_degree: ExtensionDegree) -> Result<Self, RangeProofError> {
+        if extension_degree as usize > RISTRETTO_NUMS_POINTS.len() ||
+            extension_degree as usize > RISTRETTO_NUMS_POINTS_COMPRESSED.len()
+        {
+            return Err(RangeProofError::ExtensionDegree(
+                "Not enough Ristretto NUMS points to construct the extended commitment factory".to_string(),
+            ));
+        }
+        let mut g_base_vec = Vec::with_capacity(extension_degree as usize);
+        g_base_vec.push(RISTRETTO_PEDERSEN_G);
+        let mut g_base_compressed_vec = Vec::with_capacity(extension_degree as usize);
+        g_base_compressed_vec.push(RISTRETTO_PEDERSEN_G_COMPRESSED);
+        for i in 1..extension_degree as usize {
+            g_base_vec.push(RISTRETTO_NUMS_POINTS[i]);
+            g_base_compressed_vec.push(RISTRETTO_NUMS_POINTS_COMPRESSED[i]);
+        }
```

Comment:
doesn't `RISTRETTO_NUMS_POINTS[..extension_degree as usize].to_vec()` do all of this in one line? Same for compressed

Reply:
It's actually a bit more complicated since G is not point 0, but it's still nice and compact:

```rust
    let g_base_vec: std::iter::once(&RISTRETTO_PEDERSEN_G).chain(RISTRETTO_NUMS_POINTS[1..n].iter()).cloned().collect();
    let g_base_compressed_vec = std::iter::once(&RISTRETTO_PEDERSEN_G_COMPRESSED).chain(RISTRETTO_NUMS_POINTS_COMPRESSED[1..n].iter()).cloned().collect();
```